modules/core/src/v2/trading/tradingAccount.ts

```diff
@@ -113,41 +118,47 @@ export class TradingAccount {
     const paramsCopy = JSON.parse(JSON.stringify(params)); // needs to be a deep copy
 
     // Verifies that for each party in the payload, we requested a matching party, only checking sensitive fields
-    let partiesMatch = true;
-    for (const party of payloadObj.otherParties) {
-      const matchingExpectedParty = paramsCopy.otherParties.findIndex(
-        expectedParty =>
-          party.accountId === expectedParty.accountId &&
-          party.currency === expectedParty.currency &&
-          party.amount === expectedParty.amount
+    let validAmounts = 0;
+    for (const amount of payloadObj.amounts) {
+      const matchingExpectedParty = paramsCopy.amounts.findIndex(
+        expectedAmount =>
+          amount.accountId === expectedAmount.accountId &&
+          amount.sendCurrency === expectedAmount.sendCurrency &&
+          amount.sendSubtotal === expectedAmount.sendAmount &&
+          amount.receiveAmount === expectedAmount.receiveAmount &&
+          amount.receiveCurrency === expectedAmount.receiveCurrency
       );
 
       if (matchingExpectedParty === -1) {
-        partiesMatch = false;
+        // matchingExpectedParty not found to the payloadObject
+        // payload is not valid
         break;
       }
 
-      // delete so we ensure no duplicates
-      paramsCopy.otherParties.splice(matchingExpectedParty, 1);
-    }
+      if (amount.fees && amount.fees.length > 0) {
+        let feeTotal = new BigNumber(0);
+        for (const fee of amount.fees) {
+          feeTotal = feeTotal.plus(new BigNumber(fee.feeAmount));
+        }
+
+        const expectedTotalAmount = new BigNumber(paramsCopy.amounts[matchingExpectedParty].sendAmount).plus(feeTotal);
+        if (expectedTotalAmount.toString() !== amount.sendAmount) {
+          // expected total does not match the sendAmount of the payload
+          // payload is not valid
+          break;
+        }
+      }
 
-    // the amount field will change if fees are present, but subtotal should always equal the requested send amount
-    let expectedAmount: string = params.amount;
-    if (payloadObj.fees) {
-      const totalFees = payloadObj.fees.reduce(
-        (fees: BigNumber, feeObj) => fees.plus(feeObj.feeAmount),
-        new BigNumber(0)
-      );
-      expectedAmount = new BigNumber(payloadObj.subtotal).plus(totalFees).toString();
+      // matching party found, and fee found
+      validAmounts = validAmounts + 1;
```

Comment:
It's a bit unsafe to check for the bad case, and if you don't find it, assume the case is valid. It's better to assume the case is unsafe, and only increment `validAmounts` in the inner most if statement where you know all your conditions have been checked.

I don't think it's an issue here now, but if the logic for determining a valid match ever changes, it could pose a problem.